pkg/util/dynamichelper/unstructuredobj.go

```diff
@@ -0,0 +1,60 @@
+package dynamichelper
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"strings"
+
+	"github.com/ghodss/yaml"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type UnstructuredObj struct {
+	obj unstructured.Unstructured
+}
+
+func NewUnstructuredObj(obj unstructured.Unstructured) *UnstructuredObj {
+	return &UnstructuredObj{obj}
+}
+
+func (o UnstructuredObj) GetObjectKind() schema.ObjectKind {
+	return o.obj.GetObjectKind()
+}
+
+func (o UnstructuredObj) DeepCopyObject() kruntime.Object {
+	if un := o.obj.DeepCopy(); un != nil {
+		out := NewUnstructuredObj(*un)
+		return out
+	}
+	return nil
+}
+
+func (o UnstructuredObj) GroupVersionKind() schema.GroupVersionKind {
+	return o.obj.GroupVersionKind()
+}
+
+func (o UnstructuredObj) GetNamespace() string {
+	return o.obj.GetNamespace()
+}
+func (o UnstructuredObj) GetName() string {
+	return o.obj.GetName()
+}
```

Comment:
to provide these common methods for the caller

Reply:
this wrapper seems unnecessary tbh. The only reason you need it here is because you wanted to use a method on the struct to decode it, but this is not even needed. I'd rather we just use the actual object instead of the wrapper

Reply:
they are also called [here](https://github.com/Azure/ARO-RP/pull/2555/files#diff-4dc999c394ba7fdd74cd881c96667ce6b8d1a88c5f2d5872c198e57708cc9f53R114) by guardrails controller

Reply:
I know, but there isn't any value added compared to using the actual `unstructured.Unstructured`

Reply:
feel like there is some misunderstanding here, the reason of introducing *UnstructuredObj* is that unstructured.Unstructured is not a *kruntime.Object* and hence cannot be put into the []kruntime.Object as a parameter for many APIs.

Reply:
```go
	GetObjectKind() schema.ObjectKind
	DeepCopyObject() Object
```
but then only these are required, not the others

Reply:
also, if you could add a comment saying why we have those methods that'd be great

Reply:
they get called outside dh in guardrails pkg, this is why the methods were added:
> err = r.dh.EnsureDeletedGVR(ctx, uns.GroupVersionKind().GroupKind().String(), uns.GetNamespace(), uns.GetName(), uns.GroupVersionKind().Version)
these methods are required by EnsureDeletedGVR()

Reply:
I have seen that, but it is still not useful. at that point you just need the `unstructured.Unstructured` object, and you can just use `append(creates,uns.Unstructure{obj})` when you need it. 

this should just return an `unstructured.Unstructured` object (with the signature becoming `func DecodeUnstructured(data []byte) (unstructed.Unstructured, error) {`)
```go
func (o *UnstructuredObj) DecodeUnstructured(data []byte) error {
	json, err := yaml.YAMLToJSON(data)
	if err != nil {
		return unstructured.Unstructured{},  err
	}
        obj := &unstructured.Unstructured{}
	err = obj.UnmarshalJSON(json)
        return obj, err
}
```

Reply:
dont quite understand what you mean, do you want to completely remove the UnstructuredObj?
if not, you have to expose two confusing structs to callers, ie, guardrails, in order to decode one thing, which is UnstructuredObj and unstructured.Unstructured, do that sounds good?

in my implementation, only UnstructuredObj is exposed, callers dont need to know anything about the unstructured.Unstructured, everything related to it is encapsulated inside the package.

Reply:
the unstructuredObject struct should just be here to implement the interface you need. the rest is not really useful, adding wrappers like this are making things more confusing in the end

Reply:
with the current code structure, I don't see a reasonable way to remove these methods while achieve the same functionality, if you already have something in your mind, can you please let me know? thanks

Reply:
I already did.

Reply:
ah, thanks, this is a bit tricky, I dont see other  languages have this kind of ### reverse inheritance